apps/emqx_durable_storage/test/emqx_ds_cache_SUITE.erl

```diff
@@ -0,0 +1,397 @@
+                    #{?snk_kind := ds_cache_stored_batch},
+                    2_000
+                ),
+            [{_Rank, Stream}] = emqx_ds:get_streams(DB, TopicFilter, NowMS),
+            {ok, Iter0} = emqx_ds:make_iterator(DB, Stream, TopicFilter, NowMS),
+            {ok, Iter1, Batch1} = iterate(DB, Iter0, 1, #{use_cache => true}),
+            %% The next fetches should use the cache.
+            {ok, _Iter2, Batch2} = iterate(DB, Iter1, 2, #{use_cache => true}),
+            Fetched = [Msg || {_DSKey, Msg} <- Batch1 ++ Batch2],
+            ?assertEqual(Msgs, Fetched),
+            ok
+        end,
+        fun(Trace) ->
+            ?assertMatch([#{batch := [_, _, _]}], ?of_kind(ds_cache_stored_batch, Trace)),
+            %% First fetch is a miss because the initial iterator doesn't have a last
+            %% seen key
+            ?assertMatch([#{last_seen_key := undefined}], ?of_kind(ds_cache_miss, Trace)),
+            ?assertMatch([_, _], ?of_kind(ds_cache_hit, Trace)),
+            ok
+        end
+    ),
+    ok.
+
+t_key_deleted_while_iterating(_Config) ->
+    DB = ?FUNCTION_NAME,
+    ?check_trace(
+        begin
+            ?force_ordering(
+                #{?snk_kind := key_deleted},
+                #{?snk_kind := ds_cache_lookup_enter}
+            ),
+
+            spawn(fun() ->
+                {ok, #{key := Key}} =
+                    ?block_until(#{?snk_kind := ds_cache_will_fetch}),
+                ?assertNotEqual('$end_of_table', Key),
+                emqx_ds_cache:delete(DB, Key),
+                ?tp(key_deleted, #{}),
+                ok
+            end),
+
+            TopicFilterBin = <<"some/+">>,
+            TopicFilter = emqx_topic:words(TopicFilterBin),
+            ok = open_db(DB, #{batch_size => 5, topic_filters => [TopicFilter]}),
+            NowMS = now_ms(),
+            Msgs =
+                [Msg1 | _] = [
+                    message(<<"some/topic">>, <<"1">>, NowMS + 60_000),
+                    message(<<"some/topic">>, <<"2">>, NowMS + 60_001),
+                    message(<<"some/topic">>, <<"2">>, NowMS + 60_002)
+                ],
+            {ok, {ok, _}} =
+                ?wait_async_action(
+                    emqx_ds:store_batch(DB, Msgs),
+                    #{?snk_kind := ds_cache_stored_batch},
+                    2_000
+                ),
+            [{_Rank, Stream}] = emqx_ds:get_streams(DB, TopicFilter, NowMS),
+            {ok, Iter0} = emqx_ds:make_iterator(DB, Stream, TopicFilter, NowMS),
+            {ok, Iter1, Batch1} = emqx_ds:next(DB, Iter0, 1, #{use_cache => false}),
+            {ok, _Iter2, Batch2} = emqx_ds:next(DB, Iter1, 2, #{use_cache => true}),
+            %% We don't report a cache miss in this case, as this case is equivalent to
+            %% when the client is up to date with the stream and it's the cache that needs
+            %% to fetch more.
+            %% TODO: what to do if the client started from an old key and it got GC'ed?
```

Comment:
actually, it'll then notice that last seen key is missing from the cache and fetch again; should test this.